mt-kahypar/datastructures/static_graph.cpp

```diff
@@ -238,58 +245,71 @@ namespace mt_kahypar::ds {
+        const size_t incident_edges_start = tmp_incident_edges_prefix_sum[coarse_node];
+        const size_t incident_edges_end = tmp_incident_edges_prefix_sum[coarse_node + 1];
+        const size_t size_of_range = incident_edges_end - incident_edges_start;
+
+        // Insert incident edges into concurrent bucket map
+        const size_t num_chunks = std::thread::hardware_concurrency();
+        const size_t chunk_size = (size_of_range + num_chunks - 1) / num_chunks;
+        tbb::parallel_for(UL(0), num_chunks, [&](const size_t chunk_id) {
+          const size_t start = std::min(incident_edges_start + chunk_id * chunk_size, incident_edges_end);
+          const size_t end = std::min(incident_edges_start + (chunk_id + 1) * chunk_size, incident_edges_end);
+          // First, we apply a deduplication step to the thread-local range. The reason is that on large irregular
+          // graphs, extremely large clusters (thousands of nodes) can be created during coarsening. In this case,
+          // there can be thousands of edges with the same target, which creates a massive imbalance and thus high
+          // contention in the bucket map if we insert them directly. The local deduplication avoids this problem
+          // by ensuring that each edge appears at most t times.
+          const HyperedgeID local_degree = deduplicate_tmp_edges(tmp_edges.data() + start, tmp_edges.data() + end);
+          for (size_t pos = start; pos < start + local_degree; ++pos) {
+            const TmpEdgeInformation& edge = tmp_edges[pos];
+            ASSERT(edge.isValid());
+            incident_edges_map.insert(edge.getTarget(), TmpEdgeInformation(edge));
           }
+        }, tbb::static_partitioner());
+
+        // Process each bucket in parallel and deduplicate the edges
+        std::atomic<size_t> incident_edges_pos(incident_edges_start);
+        tbb::parallel_for(UL(0), incident_edges_map.numBuckets(), [&](const size_t bucket) {
+          auto& incident_edges_bucket = incident_edges_map.getBucket(bucket);
+          const HyperedgeID bucket_degree = deduplicate_tmp_edges(incident_edges_bucket.data(),
+                                              incident_edges_bucket.data() + incident_edges_bucket.size());
+          const size_t tmp_incident_edges_pos = incident_edges_pos.fetch_add(bucket_degree);
+          memcpy(tmp_edges.data() + tmp_incident_edges_pos,
+                 incident_edges_bucket.data(), sizeof(TmpEdgeInformation) * bucket_degree);
+          incident_edges_map.clear(bucket);
         });
 
-        const size_t contracted_size = incident_edges_pos.total_sum();
+        const size_t contracted_size = incident_edges_pos.load() - incident_edges_start;
         node_sizes[coarse_node] = contracted_size;
+        resulting_ranges.emplace_back(incident_edges_start, incident_edges_pos.load());
       }
+
+      // We still sort the adjacent edges since this results in better cache locality when accessing the neighbors
```

Comment:
Please add a comment that this is also necessary for determinism. It would be a good idea to document determinism pitfalls in general. Future changes could then be done with less scrutiny.